src/parquet/util/crypto.cc

```diff
@@ -0,0 +1,386 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <memory>
+#include "parquet/util/crypto.h"
+#include <openssl/aes.h>
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+#include <string>
+#include <sstream>
+#include <iostream>
+#include "parquet/exception.h"
+
+namespace parquet {
+
+constexpr int gcm_tag_len = 16;
+constexpr int gcm_iv_len = 12;
+constexpr int ctr_iv_len = 16;
+constexpr long max_bytes = 32;
+
+struct EvpCipherCtxDeleter {
+  void operator()(EVP_CIPHER_CTX *ctx) const {
+    if (nullptr != ctx) { 
+      EVP_CIPHER_CTX_free(ctx);
+   }
+  }
+};
+
+using EvpCipherCtxPtr = std::unique_ptr<EVP_CIPHER_CTX, EvpCipherCtxDeleter>;
```

Comment:
Why not use simply define a class/struct with a destructor? Why is `std::unique_ptr` necessary?

Reply:
Suggested by @majetideepak and @jamesclampffer . Let me know guys which one should I use.

Reply:
I think this is overly elaborate. Unless you expect to need to transfer ownership of a smart pointer to some other call frame, using RAII with a simple class with a destructor that frees the resource is the simplest solution

Reply:
I usually prefer the `std::unique_ptr` approach to avoid creating a new class. `std::unique_ptr` follows the RAII design principle. I am OK with either RAII approaches.

Reply:
Well the deleter has already created a new class. Which of these do you prefer?

```
struct EvpCipherCtxDeleter {
  void operator()(EVP_CIPHER_CTX *ctx) const {
    if (nullptr != ctx) {
      EVP_CIPHER_CTX_free(ctx);
   }
  }
};

using EvpCipherCtxPtr = std::unique_ptr<EVP_CIPHER_CTX, EvpCipherCtxDeleter>;

int gcm_encrypt(const uint8_t *plaintext, int plaintext_len,
                   uint8_t *key, int key_len, uint8_t *aad, int aad_len,
                   uint8_t *ciphertext)
{
  int len;
  int ciphertext_len;
  uint8_t tag[gcm_tag_len];
  uint8_t iv[gcm_iv_len];

  // Random IV
  RAND_load_file("/dev/urandom", max_bytes);
  RAND_bytes(iv, sizeof(iv));

  // Init cipher context
  EvpCipherCtxPtr ctx(EVP_CIPHER_CTX_new());
  if(nullptr == ctx.get()) {
    throw ParquetException("Couldn't init cipher context");
  }

  if (16 == key_len) {
    // Init AES-GCM with 128-bit key
    if(1 != EVP_EncryptInit_ex(ctx.get(), EVP_aes_128_gcm(), nullptr, nullptr, nullptr)) {
      throw ParquetException("Couldn't init AES_GCM_128 encryption");
    }
  }
  else if (24 == key_len) {
    // Init AES-GCM with 192-bit key
    if(1 != EVP_EncryptInit_ex(ctx.get(), EVP_aes_192_gcm(), nullptr, nullptr, nullptr)) {
      throw ParquetException("Couldn't init AES_GCM_192 encryption");
    }
  }
  else if (32 == key_len) {
    // Init AES-GCM with 256-bit key
    if(1 != EVP_EncryptInit_ex(ctx.get(), EVP_aes_256_gcm(), nullptr, nullptr, nullptr)) {
      throw ParquetException("Couldn't init AES_GCM_256 encryption");
    }
  }

  // Setting key and IV
  if(1 != EVP_EncryptInit_ex(ctx.get(), nullptr, nullptr, key, iv)) {
    throw ParquetException("Couldn't set key and IV");
  }
  ...
```

or

```
class EvpCipher {
 public:
  explicit EvpCipher(int key_len) {
    ctx_ = EVP_CIPHER_CTX_new();
    if(nullptr == ctx_) {
      throw ParquetException("Couldn't init cipher context");
    }

    Init(key_len);
  }

  ~EvpCiphper() {
    if (nullptr != ctx_) {
      EVP_CIPHER_CTX_free(ctx_);
   }
  }

  void Init(const key_len) {
    if (16 == key_len) {
      // Init AES-GCM with 128-bit key
      if(1 != EVP_EncryptInit_ex(ctx.get(), EVP_aes_128_gcm(), nullptr, nullptr, nullptr)) {
        throw ParquetException("Couldn't init AES_GCM_128 encryption");
      }
    }
    else if (24 == key_len) {
      // Init AES-GCM with 192-bit key
      if(1 != EVP_EncryptInit_ex(ctx.get(), EVP_aes_192_gcm(), nullptr, nullptr, nullptr)) {
        throw ParquetException("Couldn't init AES_GCM_192 encryption");
      }
    }
    else if (32 == key_len) {
      // Init AES-GCM with 256-bit key
      if(1 != EVP_EncryptInit_ex(ctx.get(), EVP_aes_256_gcm(), nullptr, nullptr, nullptr)) {
        throw ParquetException("Couldn't init AES_GCM_256 encryption");
      }
    }

    // Setting key and IV
    if(1 != EVP_EncryptInit_ex(ctx.get(), nullptr, nullptr, key, iv)) {
      throw ParquetException("Couldn't set key and IV");
    }
  }

 private:
  EVP_CIPHER_CTX ctx_;
};

int gcm_encrypt(const uint8_t *plaintext, int plaintext_len,
                   uint8_t *key, int key_len, uint8_t *aad, int aad_len,
                   uint8_t *ciphertext)
{
  int len;
  int ciphertext_len;
  uint8_t tag[gcm_tag_len];
  uint8_t iv[gcm_iv_len];

  // Random IV
  RAND_load_file("/dev/urandom", max_bytes);
  RAND_bytes(iv, sizeof(iv));

  EvpCipher cipher(key_len);
  ...
```

I think that encapsulation is a good idea

Reply:
@wesm , @majetideepak Thank you for the comments and suggestions. I'm posting a new commit that takes them into account, let me know if additional changes are required.